ci: Use environment variables to set the major and minor versions

martin-g opened the pull request:
Use same versions as in the other Docker images:
- major -> 3
- minor -> 0

Use `docker manifest` cli to create and push the manifest.
Improve the step that installs QEMU to be better prepared for adding support for more CPU architectures.

Suggested-at: https://github.com/bioconda/bioconda-containers/pull/68#issuecomment-1893793704

martin-g committed:
Use environment variables to set the major and minor versions

Use same versions as in the other Docker images:
- major -> 3
- minor -> 0

Suggested-at: https://github.com/bioconda/bioconda-containers/pull/68#issuecomment-1893793704

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
@daler Please review when you have a minute!

daler commented:
Thanks. At first glance looks good, but I'm trying to figure out why we get the skipped job that looks like "Build image / quay.io/bioconda/${{ matrix.cfg.DOCKER_MANIFEST }}:${{ matrix.cfg.DOCKER_TAG }}":

![image](https://github.com/bioconda/bioconda-containers/assets/115406/97f768de-d7b5-4c85-ac87-dfb47aa70166)

I would have expected the placeholders to be filled in. Looking at the earlier jobs in the yaml, they're not using `${{ matrix.include.image }}` but rather `${{ matrix.image }}`. Maybe the `cfg:` part is left over from a different CI system and not needed?

This is hard to test because of the jobs that only run on master. i think the best way to do test this is to edit the CI yaml to upload on a mock branch ("mock-master" or something) and also edit the names and containers to include "test" or something so they can be easily deleted from the registry later. Then merge into the "mock-master" branch to verify that upload doesn't happen here, but does happen upon merge to mock-master. Once we confirm those containers and manifests are correctly uploaded on the mock branch, then branches and image names can be reverted back to the production values.

martin-g commented:
Actually I have tried to debug this few weeks ago. 
Using `include` (it has special [meaning](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations))  or `cfg` (or any other custom value) does not change the behavior.

I will try to debug it again in my fork! If I don't find a way I will ask in Github Actions support forums.

Yikun commented:
> Using include (it has special [meaning](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations)) or cfg (or any other custom value) does not change the behavior.

If we are using `include`, then:
```yaml
matrix:
  include:
    - DOCKER_MANIFEST: bioconda-utils-build-env-cos7
      DOCKER_TAG: "latest"
      DOCKER_IMAGES: "quay.io/<<USER>>/bioconda-utils-build-env-cos7:<<TAG>>,quay.io/<<USER>>/bioconda-utils-build-env-cos7-aarch64:<<TAG>>"
```

Then the `${{ matrix.DOCKER_IMAGES }}` should be used but not  `${{ matrix.include.DOCKER_IMAGES }}`.

And if we are using `cfg` as 1st level key, I am not sure 2nd level KVs has been supported by github action. [1]

I we don't want to use matrix, maybe `output` also a good choice in here? [2]

[1] https://stackoverflow.com/a/68940067
[2] https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idoutputs

martin-g commented:
It looks like a Github Actions issue to me.

A PR with a new dummy workflow that I created just to debug the issue: https://github.com/martin-g/bioconda-containers/pull/8
A run with **disabled** job - the job name is **not** interpolated: https://github.com/martin-g/bioconda-containers/actions/runs/7622228443
A run with **enabled** job - the job name **is** interpolated: https://github.com/martin-g/bioconda-containers/actions/runs/7622235752/job/20759944758

I'll keep you posted!

daler commented:
Ah, interesting.

In looking a little more closely at this, my interpretation is that

```yaml
matrix:
  include:
    - DOCKER_MANIFEST: bioconda-utils-build-env-cos7
      DOCKER_TAG: "latest"
      DOCKER_IMAGES: "quay.io/<<USER>>/bioconda-utils-build-env-cos7:<<TAG>>,quay.io/<<USER>>/bioconda-utils-build-env-cos7-aarch64:<<TAG>>"
```

only needs to be in the matrix so that the later job using the `Noelware/docker-manifest-action@v0.3.0` action can access things.

I'm always a little concerned using 3rd-party actions. Sure, they're convenient, but if anything goes wrong it's that much more of a headache to figure out and fix.

Can we instead just use bash? I haven't learned the details regarding manifests, but would something like this work with [`docker manifest`](https://docs.docker.com/engine/reference/commandline/manifest/#create-and-push-a-manifest-list)?

```bash

docker manifest create \
  quay.io/bioconda/bioconda-utils-build-env-cos7:${{ env.MAJOR_VERSION}.${{ env.MINOR_VERSION }} \
  quay.io/bioconda/bioconda-utils-build-env-cos7-x86_64:${{ env.MAJOR_VERSION }}.${{ env.MINOR_VERSION }} \
  quay.io/bioconda/bioconda-utils-build-env-cos7-aarch64:${{ env.MAJOR_VERSION }}.${{ env.MINOR_VERSION }}

# maybe some docker manifest annotate calls here?

docker manifest inspect \
  quay.io/bioconda/bioconda-utils-build-env-cos7:${{ env.MAJOR_VERSION}.${{ env.MINOR_VERSION }}

docker manifest push \
  quay.io/bioconda/bioconda-utils-build-env-cos7:${{ env.MAJOR_VERSION }}.${{ env.MINOR_VERSION }}

```

if so, that would bypass the need for the matrix stuff and would be way more maintainable over the long term. I may be missing something here though.

martin-g commented:
I've used a matrix because I anticipated adding more tags in addition to `latest`, e.g. `${MAJOR_VER}.${MINOR_VER}` and `${MAJOR_VER}`. See https://github.com/martin-g/bioconda-containers/pull/9

No matter whether `Noelware/docker-manifest-action` is used or not the [issue](https://github.com/orgs/community/discussions/101546) with the non-interpolated job name will be still there. I will simplify the job name to not use variables until this issue is fixed!

A successful Docker manifest push could be seen in my fork (uses `Noelware/docker-manifest-action`):
- PR: https://github.com/martin-g/bioconda-containers/pull/3
- run: https://github.com/martin-g/bioconda-containers/actions/runs/7087218329/job/19287324864?pr=3

But I will rework it tomorrow to use `docker manifest ***` commands!

martin-g committed:
Move the environment variables to be workflow level

Install QEMU for all CPU architectures but amd64. At the moment this is
just arm64 but now adding a new arch will not require changes to this
step

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
Closing this PR to be able to test the branch in my fork.

martin-g committed:
Build and push manifests for several tags

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Temporary enable pushing

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Use vars. context for the Quay.io username

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Convert Bash array 'tags' to JSON array

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Debug bash array to json array

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
More debug

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Convert Bash string to JSON array.

Split by whitespace

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Use `docker manifest` cli instead of Noelware/docker-manifest-action

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
Here is the PR with successful build + push to my Quay.io account:
- PR: https://github.com/martin-g/bioconda-containers/pull/10
- run: https://github.com/martin-g/bioconda-containers/actions/runs/7638262506/job/20809089060
- Quay.io
     - manifest: https://quay.io/repository/mgrigorov/bioconda-utils-build-env-cos7?tab=tags   
     - x86_64: https://quay.io/repository/mgrigorov/bioconda-utils-build-env-cos7-x86_64?tab=tags
     - aarch64: https://quay.io/repository/mgrigorov/bioconda-utils-build-env-cos7-aarch64?tab=tags

martin-g committed:
Go back to `secrets.QUAY_BIOCONDA_USERNAME`

Remove debug statements and commented out code

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Re-enable the check for push to main before pushing the manifest

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>
